# Logistic Regression on Iris Data Set Using the Artificial Neural Network

## Run ANN logistic regression on the Iris Train data

In [1]:
# Import numpy
import numpy as np

# Unity/Identity activation function for layer 0
def identity_activation(G0):
    return G0

# TanH activation function for layer 1
def tanh_activation(G1):
    return (np.exp(G1)-np.exp(-G1))/(np.exp(G1)+np.exp(-G1))

# Sigmoid activation function for layer 2
def sigmoid_activation(G2):
    return 1/(1+np.exp(-G2))

# Initialize Weights and Biases
def initialize_weights_and_biases(n):
    W0 = np.identity(n)
    b0 = np.zeros((n,1))
    
    W1 = np.identity(n)
    b1 = np.zeros((n,1))
    
    W2 = np.ones((1,n))
    b2 = np.ones((1,1))
    
    return W0, b0, W1, b1, W2, b2

# Cost function
def cost_function(Y, Yhat, m):
    return -(1/m)*np.sum(Y*np.log(Yhat) + (1-Y)*np.log(1-Yhat))

def ann_logit(filename):
    
    # Read CSV data from input file
    file_csv = np.genfromtxt (filename, delimiter=",")

    X = file_csv[1:,0:4].transpose()
    Y = file_csv[1:,4:5].transpose()

    # Number of training examples
    m = X.shape[1]

    # Number of features
    n = X.shape[0]

    # Iterations
    h = 100000

    # Learning rate
    alpha = 0.05

    # Initialize the Weights and Biases
    W0, b0, W1, b1, W2, b2 = initialize_weights_and_biases(n)

    # Initialize the J cost function array
    J = np.zeros((1,h))

    # Loop through the forward and backward steps
    for g in range(h):

        # Forward Propagation through the layers 0, 1 and 2
        # Layer 0 - Unity
        X1 = identity_activation(W0@X + b0)

        # Layer 1 - Unity
        X2 = identity_activation(W1@X1 + b1)

        # Layer 2 - Sigmoid
        Yhat = sigmoid_activation(W2@X2 + b2)

        # Calculate the Cost Function
        J[0][g] = cost_function(Y, Yhat, m)

        # Use derivatives of J cost funtion to update the Weights and Biases
        dJdz = Yhat - Y
        dJdb = (1/m)*np.sum(dJdz)
        dJdw = (1/m)*dJdz@np.transpose(X2)
        b2 = b2 - alpha*dJdb
        W2 = W2 - alpha*dJdw

    print("The Weights array after training is: {}".format(W2))
    print("The Biases array after training is: {}".format(b2))
    print("The last value of the Cost Function is: {}".format(J[0][h-1]))

# Call the ANN logitistic regression for the training data set
ann_logit('iristrain-1.csv')


The Weights array after training is: [[-1.37203655 -3.55865762  5.82179189  3.3445902 ]]
The Biases array after training is: [[0.08939374]]
The last value of the Cost Function is: 0.00023010959719113575


The Weights array after training is: [[-1.37203655 -3.55865762  5.82179189  3.3445902 ]]

The Biases array after training is: [[0.08939374]]

The last value of the Cost Function is: 0.00023010959719113575

## Run ANN logistic regression on the Iris Test data

In [2]:

# Call the ANN logitistic regression for the testing data set
ann_logit('iristest-1.csv')


The Weights array after training is: [[-1.5575962  -3.09156747  5.47374584  3.04684905]]
The Biases array after training is: [[0.08493814]]
The last value of the Cost Function is: 0.00014143460554265697


The Weights array after training is: [[-1.5575962  -3.09156747  5.47374584  3.04684905]]

The Biases array after training is: [[0.08493814]]

The last value of the Cost Function is: 0.00014143460554265697